qiskit_nature/second_q/operators/spin_op.py

```diff
@@ -18,532 +18,582 @@
+
+
+    **Iteration**
+
+    Instances of `SpinOp` are iterable. Iterating a SpinOp yields (term, coefficient)
+    pairs describing the terms contained in the operator. Labels containing powers/exponents
+    will be expanded into multiple (term, coefficient) pairs.
 
+    Attributes:
+        num_spins: the number of spins on which this operator acts. This is
+            considered a lower bound, which means that mathematical operations acting on two or more
+            operators will result in a new operator with the maximum number of spins of any
+            of the involved operators.
     """
 
+    _OPERATION_REGEX = re.compile(r"([XYZ]_\d+(\^\d+)?\s)*[XYZ]_\d+(\^\d+)?")
+
     def __init__(
         self,
-        data: Union[
-            str,
-            List[Tuple[str, complex]],
-            Tuple[np.ndarray, np.ndarray],
-        ],
-        spin: Union[float, Fraction] = Fraction(1, 2),
-        register_length: Optional[int] = None,
+        data: Mapping[str, _TCoeff],
+        spin: float | Fraction = Fraction(1, 2),  # TODO: extend to list
+        num_spins: int | None = None,
+        *,
+        copy: bool = True,
+        validate: bool = True,
     ):
         r"""
         Args:
             data: label string, list of labels and coefficients. See the label section in
                   the documentation of :class:`SpinOp` for more details.
             spin: positive half-integer (integer or half-odd-integer) that represents spin.
-            register_length: length of the particle register.
+            num_spins: the number spins on which this operator acts.
+            copy: when set to False the `data` will not be copied and the dictionary will be
+                stored by reference rather than by value (which is the default; `copy=True`). Note,
+                that this requires you to not change the contents of the dictionary after
+                constructing the operator. This also implies `validate=False`. Use with care!
+            validate: when set to False the `data` keys will not be validated. Note, that the
+                SparseLabelOp base class, makes no assumption about the data keys, so will not
+                perform any validation by itself. Only concrete subclasses are encouraged to
+                implement a key validation method. Disable this setting with care!
 
         Raises:
-            ValueError: invalid data is given.
-            QiskitNatureError: invalid spin value
+            QiskitNatureError: when an invalid key is encountered during validation.
+            QiskitNatureError: when spin is not a positive half-integer.
         """
-        self._coeffs: np.ndarray
-        self._spin_array: np.ndarray
-        dtype = np.complex128  # TODO: configurable data type. mixin?
-
+        self.num_spins = num_spins
         spin = Fraction(spin)
-        if spin.denominator not in (1, 2):
+        if spin.denominator not in (1, 2) or spin < 0:
```

Comment:
```suggestion
        if spin.denominator not in (1, 2) or spin <= 0:
```
Spin 0 should not be allowed